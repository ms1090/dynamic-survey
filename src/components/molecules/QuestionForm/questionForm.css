.question-form:after {
  clear: both;
  content: "";
  display: table;
}
.question-form .question-container {
  color: #285175;
  font-weight: bold;
  margin-bottom: 10px;
}
.question-form .question-container span:last-child {
  margin-left: 5px;
}
.input-field-container {
  min-height: 240px;
}
.float-right {
  float: right;
}
