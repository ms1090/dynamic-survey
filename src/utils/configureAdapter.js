import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// configure adapter to use enzyme in unit test cases
configure({ adapter: new Adapter() });
