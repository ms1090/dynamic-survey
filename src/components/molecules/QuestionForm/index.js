import React, { Component } from 'react';
import PropTypes from 'prop-types';
import RadioBox from '../../atoms/RadioBox';
import SelectBox from '../../atoms/SelectBox';
import TextArea from '../../atoms/TextArea';
import Divider from '../../atoms/Divider';
import Button from '../../atoms/Button';
import './questionForm.css';

// Display survey question form
export default class QuestionForm extends Component {
  // Dispatch action to store user answer in state
  componentDidMount() {
    const {
      question,
      storeUserAnswer,
    } = this.props;
    if (question) {
      storeUserAnswer({ userAnswer: question.answer || '' });
    }
  }

  // Dispatch action to store user answer if updated
  componentDidUpdate(prevProps) {
    const {
      question,
      storeUserAnswer,
    } = this.props;
    if (prevProps.question.answer !== question.answer) {
      storeUserAnswer({ userAnswer: question.answer || '' });
    }
  }

  /**
   * Store user answer in state
   * @param {Object} event Event object containing user answer
   */
  handleChange = (event) => {
    const userAnswer = event.target.value;
    const { storeUserAnswer } = this.props;
    storeUserAnswer({ userAnswer });
  }

  /**
   * Set user answer, update current question index
   * and navigate to next question
   * @param {String} navDir To show next or prev question
   */
  handleNav = (navDir) => {
    const {
      setUserAnswer,
      currQuesIndex,
      quesCount,
      setCurrQuesIndex,
      resetUserAnswer,
      userAnswer,
      question: { id: currQuesId },
    } = this.props;
    let newQuesIndex;
    if (navDir === 'next') {
      if (currQuesIndex < quesCount) {
        newQuesIndex = currQuesIndex + 1;
      }
    } else if (currQuesIndex > 0) {
      newQuesIndex = currQuesIndex - 1;
    }
    const payload = {
      currQuesIndex: newQuesIndex,
      currQuesId,
      userAnswer,
    };
    setUserAnswer(payload);
    setCurrQuesIndex(newQuesIndex);
    resetUserAnswer();
  }

  // Handle "NEXT" button click event
  handleNext = () => {
    this.handleNav('next');
  }

  // Handle "BACK" button click event
  handleBack = () => {
    this.handleNav('back');
  }

  // Handle "SUBMIT" button click event
  handleSubmitSurvey = () => {
    this.handleNav('next');
  }

  /**
   * @return {React.Element} Render survey question form component UI
   */
  render() {
    const {
      question,
      userAnswer: value,
      currQuesIndex,
      quesCount,
    } = this.props;
    let userInput;
    // Create input element for survey question
    if (question.options) {
      if (question.options.length <= 5) {
        // use radio button if there are <= 5 options
        userInput = question.options.map((ele, index) => (
          <RadioBox
            name="userAns"
            id={`userAns${index}`}
            key={ele}
            value={ele}
            label={ele}
            onChange={this.handleChange}
            checked={value === ele ? 'checked' : ''}
          />
        ));
      } else {
        // use select box if there are more than 5 options
        userInput = (
          <SelectBox
            id="userAns"
            options={question.options.map(ele => ({
              value: ele, label: ele,
            }))}
            label="Select an option"
            onChange={this.handleChange}
            value={value}
          />
        );
      }
    } else {
      // use text input field if there are no options for question
      userInput = (
        <TextArea
          id="userAns"
          label="Your answer"
          onChange={this.handleChange}
          value={value}
        />
      );
    }
    return (
      <section className="question-form">
        <div className="question-container">
          <span>{`Question ${currQuesIndex + 1} :`}</span>
          <span>{question.text}</span>
        </div>
        <div className="input-field-container">{userInput}</div>
        <Divider />
        {currQuesIndex < quesCount - 1
          ? (
            <Button
              className="float-right"
              disabled={value ? '' : 'disabled'}
              onClick={this.handleNext}
            >
              NEXT
            </Button>
          )
          : (
            <Button
              className="float-right"
              disabled={value ? '' : 'disabled'}
              onClick={this.handleSubmitSurvey}
            >
              SUBMIT
            </Button>
          )
        }
        {currQuesIndex > 0
          ? <Button onClick={this.handleBack}>BACK</Button>
          : null
        }
      </section>
    );
  }
}
// default values for component props
QuestionForm.defaultProps = {
  quesCount: 0,
  currQuesIndex: 0,
  question: {},
  userAnswer: '',
  setUserAnswer: null,
  resetUserAnswer: null,
  storeUserAnswer: null,
  setCurrQuesIndex: null,
};
// prop types checking
QuestionForm.propTypes = {
  quesCount: PropTypes.number,
  currQuesIndex: PropTypes.number,
  question: PropTypes.shape({
    id: PropTypes.string,
    options: PropTypes.arrayOf(PropTypes.string),
    text: PropTypes.string,
    answer: PropTypes.string,
  }),
  userAnswer: PropTypes.string,
  setUserAnswer: PropTypes.func,
  resetUserAnswer: PropTypes.func,
  storeUserAnswer: PropTypes.func,
  setCurrQuesIndex: PropTypes.func,
};
