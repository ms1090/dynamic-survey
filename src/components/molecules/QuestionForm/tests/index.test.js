import React from 'react';
import { shallow } from 'enzyme';
import QuestionForm from '..';
import toJson from 'enzyme-to-json';
import 'utils/configureAdapter';

describe('QuestionForm Component', () => {
  const props = {
    quesCount: 4,
    currQuesIndex: 1,
    question: {
      id: 'ques2',
      text: 'Can you tell us why you have this view?',
    },
    userAnswer: 'test answer',
    setUserAnswer: () => {},
    resetUserAnswer: () => {},
    storeUserAnswer: () => {},
    setCurrQuesIndex: () => {},
  };
  test('should render correctly', () => {
    let wrapper = shallow(<QuestionForm {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper = shallow(<QuestionForm
      {...props}
      currQuesIndex={4}
      question={
        {
          id: 'ques1',
          text: 'What do you think about the use of dummies?',
          options: [
            "It's fine- whatever works, right?",
            "I don't mind them but never used much.",
            'I preferred to not use the dummy as much as possible.',
            'I would never use a dummy.',
          ],
        }
      }
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper = shallow(<QuestionForm
      {...props}
      currQuesIndex={0}
      question={
        {
          id: 'ques3',
          text: 'At what age do you think it is appropriate for a child to have a dummy?',
          options: [
            'Up to 1 yr',
            'Up to 2 yrs',
            'Up to 3 yrs',
            'Up to 4yrs',
            'Up to 5yrs',
            'Beyond 5yrs',
          ],
        }
      }
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  test('should call storeUserAnswer on mount', () => {
    const storeUserAnswerMock = jest.fn();
    shallow(<QuestionForm
      {...props}
      storeUserAnswer={storeUserAnswerMock}
    />);
    expect(storeUserAnswerMock.mock.calls.length).toBe(1);
  });
  test('should call storeUserAnswer on change event', () => {
    const storeUserAnswerMock = jest.fn();
    const wrapper = shallow(<QuestionForm
      {...props}
      storeUserAnswer={storeUserAnswerMock}
    />);
    wrapper.find('TextArea').simulate('change', { target: { value: 'test data' } });
    expect(storeUserAnswerMock.mock.calls.length).toBe(2);
  });
  test(
    'should navigate to next/prev question on click of "NEXT"/"BACK" button',
    () => {
      const mocks = {
        setUserAnswer: jest.fn(),
        setCurrQuesIndex: jest.fn(),
        resetUserAnswer: jest.fn(),
      };
      const wrapper = shallow(<QuestionForm
        {...props}
        {...mocks}
      />);
      const buttonEles = wrapper.find('Button');
      buttonEles.at(0).simulate('click');
      buttonEles.at(1).simulate('click');
      expect(mocks.setUserAnswer.mock.calls.length).toBe(2);
      expect(mocks.setCurrQuesIndex.mock.calls.length).toBe(2);
      expect(mocks.resetUserAnswer.mock.calls.length).toBe(2);
    },
  );
  test('should complete survey on click of "SUBMIT" button', () => {
    const mocks = {
      setUserAnswer: jest.fn(),
      setCurrQuesIndex: jest.fn(),
      resetUserAnswer: jest.fn(),
    };
    const wrapper = shallow(<QuestionForm
      {...props}
      {...mocks}
      currQuesIndex={4}
    />);
    wrapper.find('Button').at(0).simulate('click');
    expect(mocks.setUserAnswer.mock.calls.length).toBe(1);
    expect(mocks.setCurrQuesIndex.mock.calls.length).toBe(1);
    expect(mocks.resetUserAnswer.mock.calls.length).toBe(1);
  });
  test('should call storeUserAnswer on update', () => {
    const storeUserAnswerMock = jest.fn();
    const wrapper = shallow(<QuestionForm
      {...props}
      storeUserAnswer={storeUserAnswerMock}
    />);
    wrapper.setProps({
      question: {
        id: 'ques2',
        text: 'Can you tell us why you have this view?',
        answer: 'test answer',
      },
    });
    expect(storeUserAnswerMock.mock.calls.length).toBe(2);
  });
});
