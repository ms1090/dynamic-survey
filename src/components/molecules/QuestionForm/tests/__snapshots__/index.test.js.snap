// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionForm Component should render correctly 1`] = `
<section
  className="question-form"
>
  <div
    className="question-container"
  >
    <span>
      Question 2 :
    </span>
    <span>
      Can you tell us why you have this view?
    </span>
  </div>
  <div
    className="input-field-container"
  >
    <TextArea
      className=""
      id="userAns"
      label="Your answer"
      onChange={[Function]}
      value="test answer"
    />
  </div>
  <Divider />
  <Button
    className="float-right"
    disabled=""
    onClick={[Function]}
    primary={false}
    secondary={false}
  >
    NEXT
  </Button>
  <Button
    className=""
    onClick={[Function]}
    primary={false}
    secondary={false}
  >
    BACK
  </Button>
</section>
`;

exports[`QuestionForm Component should render correctly 2`] = `
<section
  className="question-form"
>
  <div
    className="question-container"
  >
    <span>
      Question 5 :
    </span>
    <span>
      What do you think about the use of dummies?
    </span>
  </div>
  <div
    className="input-field-container"
  >
    <RadioBox
      checked=""
      className=""
      id="userAns0"
      key="It's fine- whatever works, right?"
      label="It's fine- whatever works, right?"
      name="userAns"
      onChange={[Function]}
      value="It's fine- whatever works, right?"
    />
    <RadioBox
      checked=""
      className=""
      id="userAns1"
      key="I don't mind them but never used much."
      label="I don't mind them but never used much."
      name="userAns"
      onChange={[Function]}
      value="I don't mind them but never used much."
    />
    <RadioBox
      checked=""
      className=""
      id="userAns2"
      key="I preferred to not use the dummy as much as possible."
      label="I preferred to not use the dummy as much as possible."
      name="userAns"
      onChange={[Function]}
      value="I preferred to not use the dummy as much as possible."
    />
    <RadioBox
      checked=""
      className=""
      id="userAns3"
      key="I would never use a dummy."
      label="I would never use a dummy."
      name="userAns"
      onChange={[Function]}
      value="I would never use a dummy."
    />
  </div>
  <Divider />
  <Button
    className="float-right"
    disabled=""
    onClick={[Function]}
    primary={false}
    secondary={false}
  >
    SUBMIT
  </Button>
  <Button
    className=""
    onClick={[Function]}
    primary={false}
    secondary={false}
  >
    BACK
  </Button>
</section>
`;

exports[`QuestionForm Component should render correctly 3`] = `
<section
  className="question-form"
>
  <div
    className="question-container"
  >
    <span>
      Question 1 :
    </span>
    <span>
      At what age do you think it is appropriate for a child to have a dummy?
    </span>
  </div>
  <div
    className="input-field-container"
  >
    <SelectBox
      className=""
      id="userAns"
      label="Select an option"
      name=""
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "Up to 1 yr",
            "value": "Up to 1 yr",
          },
          Object {
            "label": "Up to 2 yrs",
            "value": "Up to 2 yrs",
          },
          Object {
            "label": "Up to 3 yrs",
            "value": "Up to 3 yrs",
          },
          Object {
            "label": "Up to 4yrs",
            "value": "Up to 4yrs",
          },
          Object {
            "label": "Up to 5yrs",
            "value": "Up to 5yrs",
          },
          Object {
            "label": "Beyond 5yrs",
            "value": "Beyond 5yrs",
          },
        ]
      }
      value="test answer"
    />
  </div>
  <Divider />
  <Button
    className="float-right"
    disabled=""
    onClick={[Function]}
    primary={false}
    secondary={false}
  >
    NEXT
  </Button>
</section>
`;
